### PR TITLE
Created function to list redirection

### PR DESCRIPTION
**Pull request by core23**
I am targeting this branch, because this is BC.

## Changelog

```markdown
### Added
- Added `CRUDController::redirectToList` method for all list redirections
```

## Subject

This is a DRY fix.

### CODE REVIEW

File: Tests/Controller/CRUDControllerTest.php [hunk 1]
```diff
@@ -969,15 +969,15 @@ public function testRedirectToWithObject()
 
         $response = $this->protectedTestedMethods['redirectTo']->invoke($this->controller, $object, $this->request);
         $this->assertInstanceOf('Symfony\Component\HttpFoundation\RedirectResponse', $response);
-        $this->assertSame('list', $response->getTargetUrl());
+        $this->assertSame('list?', $response->getTargetUrl());
```
Comment (Soullivaneuh) [inline Tests/Controller/CRUDControllerTest.php]: This should have `?` if no parameter.

Comment (core23) [reply Tests/Controller/CRUDControllerTest.php]: This is a problem of the symfony `Router`

Comment (Soullivaneuh) [reply Tests/Controller/CRUDControllerTest.php]: What do you mean? It currently works on the current branch, we can't merge this. :-/

File: Controller/CRUDController.php [hunk 2]
```diff
@@ -1129,13 +1121,25 @@ protected function redirectTo($object)
...
 
     /**
+     * Redirect the user depend on the list view.
+     *
+     * @return RedirectResponse
+     */
+    final protected function redirectToList()
+    {
+        return new RedirectResponse($this->admin->generateUrl('list', [
+            'filter' => $this->admin->getFilterParameters(),
```
Comment (Soullivaneuh) [inline Controller/CRUDController.php]: What about not passing `filter` if the parameters are empty?

Ref: https://github.com/sonata-project/SonataAdminBundle/pull/4741/files#r148931077

Comment (core23) [reply Controller/CRUDController.php]: Fixed that

File: Controller/CRUDController.php [hunk 3]
```diff
@@ -1129,13 +1121,29 @@ protected function redirectTo($object)
...
         if (!$url) {
-            $url = $this->admin->generateUrl('list');
+            return $this->redirectToList();
         }
 
         return new RedirectResponse($url);
     }
 
     /**
+     * Redirect the user depend on the list view.
```
Comment (greg0ire) [inline Controller/CRUDController.php]: What is this supposed to mean?

**Comment by greg0ire** [comment]
> This is a problem of the symfony Router

Can you please elaborate? why does `getTargetUrl` not look like an url? Are we mocking that somewhere?

**Comment by core23** [comment]
> Can you please elaborate? why does `getTargetUrl` not look like an url? Are we mocking that somewhere?

We are using a mock for this: https://github.com/core23/SonataAdminBundle/blob/d382cc0a357908c649bc98d80575e8214b9ac0c6/Tests/Controller/CRUDControllerTest.php#L433-L461

Looks like the current mock was the problem, but this final solution should work

**Comment by greg0ire** [comment]
Ok then... looks like our problem is because we mock what we don't own.

File: Controller/CRUDController.php [hunk 4]
```diff
@@ -1129,13 +1121,29 @@ protected function redirectTo($object)
...
+     */
+    final protected function redirectToList()
+    {
+        $parameter = [];
+
+        if ($filter = $this->admin->getFilterParameters()) {
+            $parameter['filter'] = $filter;
+        }
+
+        return new RedirectResponse($this->admin->generateUrl('list', $parameter));
```
Comment (Soullivaneuh) [inline Controller/CRUDController.php]: You can use `$this->redirect` here.

Comment (core23) [reply Controller/CRUDController.php]: Fixed

File: Controller/CRUDController.php [hunk 5]
```diff
@@ -1129,13 +1121,29 @@ protected function redirectTo($object)
...
         if (!$url) {
-            $url = $this->admin->generateUrl('list');
+            return $this->redirectToList();
         }
 
         return new RedirectResponse($url);
     }
 
     /**
+     * Redirect the user to the list view.
```
Comment (Soullivaneuh) [inline Controller/CRUDController.php]: Redirect**s**.

The method redirect**s**. :wink:

Comment (core23) [reply Controller/CRUDController.php]: Copy & paste typo ;)

Comment (greg0ire) [reply Controller/CRUDController.php]: Could be imperative, in which case, no s :P

Comment (Soullivaneuh) [reply Controller/CRUDController.php]: Yeah but we don't use imperative in SymfonyLand. :stuck_out_tongue:

File: Controller/CRUDController.php [hunk 6]
```diff
@@ -1087,7 +1079,7 @@ protected function handleModelManagerException(\Exception $e)
     }
 
     /**
-     * Redirect the user depend on this choice.
+     * Redirects the user depend on this choice.
```
Comment (greg0ire) [inline Controller/CRUDController.php]: I still don't understand what this comment is supposed to mean.

Comment (core23) [reply Controller/CRUDController.php]: Just fixed the grammar. Any better description @greg0ire ?

Comment (greg0ire) [reply Controller/CRUDController.php]: "redirects the user depending on his choice"? what does "this" refer to in your sentence?

File: src/Controller/CRUDController.php [hunk 7]
```diff
@@ -1142,12 +1130,28 @@ protected function redirectTo($object)
...
     }
 
+    /**
+     * Redirects the user to the list view.
+     *
+     * @return RedirectResponse
+     */
+    final protected function redirectToList()
+    {
+        $parameter = [];
```
Comment (greg0ire) [inline src/Controller/CRUDController.php]: Did you really mean to make this variable singular?

Comment (core23) [reply src/Controller/CRUDController.php]: Fixed

**Comment by greg0ire** [comment]
Thanks @core23 !

**Comment by peter-gribanov** [comment]
In vain you made the method final.

**Comment by peter-gribanov** [comment]
I used the exact same method in my project and redefined its behavior in some controllers.
After the upgrade, everything breaks.
Now i have to come up with another name for the method that does the same.

**Comment by greg0ire** [comment]
Wow bad luck! Sorry for you!

**Comment by Koc** [comment]
🤣 have same issue like @peter-gribanov

**Comment by core23** [comment]
Must be a good name for a method :trollface:

**Pull request merged**